### ISSUE TITLE
[Tags] Adapt language of DiscussionTaggedPost if tags are invisible/deleted

### ISSUE THREAD

**Issue by tobscure** [flarum/core#1143]
![fixthis](https://cloud.githubusercontent.com/assets/128862/23594541/bf9719bc-026a-11e7-8697-74acc37d7a44.png)

ie, this should just be "Toby added the Dev > Contributors-only tags."

**Comment by franzliedke** [comment]
@tobscure How should we handle a post like this?

> Franz removed the Deleted tag

**Comment by dav-is** [comment]
Wouldn't it be `Toby restored this discussion`? Those tags were already on there before it was deleted right?

**Comment by tobscure** [comment]
Nope this is referring to "Deleted" as in the tag itself has been deleted since it was added/removed.

**Comment by franzliedke** [comment]
@tobscure Do you have a comment on my question? Filtering out deleted tags is easy, but what do we do when, after filtering, nothing to display is left over? Is there an easy way to not display the event post at all - from within the event post?

**Comment by tobscure** [comment]
@franzliedke sorry, missed that, good point. I guess the ideal solution would be to filter out these posts (if they contain no tags that are visible to the user) in the API response... but that might be tough both from an extension API point of view, and also performance. So failing that, we probably need to add an API to the Post component which allows the Post to tell the PostStream "don't render me!"

**Comment by Luceos** [comment]
How about approaching this from the other side? Listen to the TagDeleted event and update/delete related EventPosts?

**Comment by tobscure** [comment]
@Luceos currently information about which tags were added/removed is stored as JSON in the post content field. As such, there's no way to look up the EventPosts by which tags they involve.

**Comment by Luceos** [comment]
- Tag gets deleted, fires an event where the Id is known.
- Listener checks posts table for type = `discussionTagged`, reads the json content
- does magic based on first array (oldTagIds) or second array (newTagIds)

What did I miss that this isn't possible?

**Comment by franzliedke** [comment]
Well, the "reading json content" part would be pretty costly, because it would need to be done in PHP. For potentially very many posts.

**Comment by tobscure** [comment]
Yeah exactly, that doesn't scale for large amounts of tags/posts.

**Comment by Luceos** [comment]
Hm right, json-type columns isn't exactly a solution either. Ok, point taken :grin:

**Comment by pflstr** [comment]
> I guess the ideal solution would be to filter out these posts (if they contain no tags that are visible to the user) in the API response... but that might be tough both from an extension API point of view, and also performance. So failing that, we probably need to add an API to the Post component which allows the Post to tell the PostStream "don't render me!"

What about deleting those event posts, which contain only tags that are not visible to the user anymore (and are not supposed to become visible again at any time in the future and are not important for other reasons)?

I do understand that it might be costly to scan the posts table for such superfluous event posts, but one could leave this check to the time, when a post stream is rendered anyway, thus telling the post stream "don't render me" and at the same time care for the deletion of the event post, so that next time the post stream is rendered, a "don't render me" isn't necessary anymore.
